### PR TITLE
Add Pooltogether protocol tests

### PR DESCRIPTION
**Pull request by alexkeating**

### CODE REVIEW

**Comment by alexkeating** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/6" target="_blank">Graphite</a> 
    * **PR #7** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/7" target="_blank">Graphite</a> 
      * **PR #9** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/9" target="_blank">Graphite</a> 
        * **PR #11** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/11" target="_blank">Graphite</a> 
          * **PR #12** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/12" target="_blank">Graphite</a> 
            * **PR #13** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/13" target="_blank">Graphite</a> 
              * **PR #14** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/14" target="_blank">Graphite</a> 
                * **PR #15** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/15" target="_blank">Graphite</a>  👈
                  * **PR #17** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/17" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/15?utm_source=stack-comment).

**Comment by alexkeating** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/6" target="_blank">Graphite</a> 
    * **PR #7** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/7" target="_blank">Graphite</a> 
      * **PR #9** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/9" target="_blank">Graphite</a> 
        * **PR #11** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/11" target="_blank">Graphite</a> 
          * **PR #12** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/12" target="_blank">Graphite</a> 
            * **PR #13** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/13" target="_blank">Graphite</a> 
              * **PR #14** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/14" target="_blank">Graphite</a> 
                * **PR #15** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/15" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/15?utm_source=stack-comment).

File: test/interfaces/IStakePrizePool.sol [hunk 1]
```diff
@@ -0,0 +1,129 @@
+//
+pragma solidity ^0.8.10;
+
+interface IStakePrizePool {
```
Comment (apbendi) [inline test/interfaces/IStakePrizePool.sol:5]: nit: might be nice to add a comment about how this interface was generated, as done for others.

File: test/PooltogetherGovernor.t.sol [hunk 2]
```diff
@@ -3,12 +3,19 @@ pragma solidity ^0.8.18;
 
 import {IERC20} from "@openzeppelin/contracts/interfaces/IERC20.sol";
 import {IGovernor} from "@openzeppelin/contracts/governance/IGovernor.sol";
+import {ERC20VotesComp} from "@openzeppelin/contracts/token/ERC20/extensions/ERC20VotesComp.sol";
```
Comment (apbendi) [inline test/PooltogetherGovernor.t.sol:6]: *arbitrary comment location*

Looks like this needs a run of the linter, some things in the test file to fix.

File: test/PooltogetherGovernor.t.sol [hunk 3]
```diff
@@ -979,3 +986,286 @@ contract CastVoteWithReasonAndParams is ProposalTest {
     }
   }
 }
+
+contract _Execute is ProposalTest {
```
Comment (apbendi) [inline test/PooltogetherGovernor.t.sol:1137]: Hmmm is the `_` required by scopelint spec?

File: test/PooltogetherGovernor.t.sol [hunk 4]
```diff
@@ -979,3 +986,286 @@ contract CastVoteWithReasonAndParams is ProposalTest {
...
+
+    uint256 safeBalance = uniswapPositionManager.balanceOf(POOLTOGETHER_UNISWAP_SAFE);
+    assertEq(safeBalance, 0);
+
+    ProposalBuilder proposals = new ProposalBuilder();
+    proposals.add(
+      UNISWAP_POSITION_CONTRACT,
+      0,
+      abi.encodeWithSignature(
+        "transferFrom(address,address,uint256)", TIMELOCK, POOLTOGETHER_UNISWAP_SAFE, 454_861
```
Comment (apbendi) [inline test/PooltogetherGovernor.t.sol]: Maybe overkill, but if it's easy enough, maybe we can use a seed and list of IDs to randomize which NFT we send?

**Review by apbendi — COMMENTED** [review]
A few initial comments @alexkeating. Looking good overall!

File: test/interfaces/IStakePrizePool.sol (hunk 1 above)
Comment (alexkeating) [reply test/interfaces/IStakePrizePool.sol:5]: https://github.com/ScopeLift/pooltogether-governor-upgrade/pull/15/commits/836515d28beed97fa04a56c882d40d8f7d0f60c0

File: test/PooltogetherGovernor.t.sol (hunk 2 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol:6]: https://github.com/ScopeLift/pooltogether-governor-upgrade/pull/15/commits/0c652d47811e76412b32af2e1a38ab52b2566ef5

File: test/PooltogetherGovernor.t.sol (hunk 3 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol:1137]: Yes, exactly the test contract name needs to match the method it tests in order for Scopelint to pick it up

File: test/PooltogetherGovernor.t.sol (hunk 4 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol]: https://github.com/ScopeLift/pooltogether-governor-upgrade/pull/15/commits/9c28b0198c35bb93c524d24ad05deff81b441db1

File: test/PooltogetherGovernor.t.sol (hunk 2 above)
Comment (apbendi) [reply test/PooltogetherGovernor.t.sol:6]: New ones showing up now 🙈

File: notes.txt [hunk 5]
```diff
@@ -0,0 +1,39 @@
+pragma solidity ^0.8.10;
```
Comment (apbendi) [inline notes.txt]: Accidentally committed?

File: test/PooltogetherGovernor.t.sol [hunk 6]
```diff
@@ -1124,3 +1131,297 @@ contract CastVoteWithReasonAndParams is ProposalTest {
...
+  }
+
+  function testFuzz_UpdateV4DripRate(uint256 newDrip) public {
+    // Drip must be greater than 0
+    // https://etherscan.io/address/0xbd537257fad96e977b9e545be583bbf7028f30b9#code#F1#L162
+    vm.assume(newDrip > 0);
+
+    IV4PooltogetherTokenFaucet tokenFaucet = IV4PooltogetherTokenFaucet(V4_TOKEN_FAUCET);
+    uint256 oldDrip = tokenFaucet.dripRatePerSecond();
+    assertEq(oldDrip, 5_787_037_037_037_037, "Old value is not correct");
```
Comment (apbendi) [inline test/PooltogetherGovernor.t.sol]: Why do we need this assertion?

File: test/PooltogetherGovernor.t.sol [hunk 7]
```diff
@@ -1124,3 +1131,297 @@ contract CastVoteWithReasonAndParams is ProposalTest {
...
+    // Assert that the drip has been changed
+    assertEq(newDrip, tokenFaucet.dripRatePerSecond(), "New drip value is not correct");
+  }
+
+  function testFuzz_UpdateV3ReserveRate(uint224 _newRate) public {
+    IV3ConfigurableReserve configurableReserve = IV3ConfigurableReserve(V3_CONFIGURABLE_RESERVE);
+
+    address reserveSource = 0x821B9819c0348076AD370b376522e1327AF7684A;
+    uint256 reserveRate = configurableReserve.reserveRateMantissa(reserveSource);
+    assertEq(reserveRate, 0, "Old reserve rate is not correct");
```
Comment (apbendi) [inline test/PooltogetherGovernor.t.sol]: Same question, why do we pull and assert against the old rate?

**Review by apbendi — COMMENTED** [review]
I know it's still in draft @alexkeating, just a few more interim comments from me.

File: notes.txt (hunk 5 above)
Comment (alexkeating) [reply notes.txt]: Yes, will remove

File: test/PooltogetherGovernor.t.sol (hunk 6 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol]: hmmm, yes I guess we could remove as it is a fuzz test.

File: test/PooltogetherGovernor.t.sol (hunk 7 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol]: Yes, I think we can remove

File: test/PooltogetherGovernor.t.sol (hunk 6 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol]: https://github.com/ScopeLift/pooltogether-governor-upgrade/pull/15/commits/6e5b21a9923b1422141be80e914489f6a954e4db

File: test/PooltogetherGovernor.t.sol (hunk 7 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol]: https://github.com/ScopeLift/pooltogether-governor-upgrade/pull/15/commits/6e5b21a9923b1422141be80e914489f6a954e4db

File: test/PooltogetherGovernor.t.sol (hunk 2 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol:6]: Should be good now

**Review by apbendi — APPROVED** [review]
Looks good @alexkeating, nice work on this!

**Pull request merged**